File: ai/src/eval_bar.py
"""
LoL Draft AI - Eval Bar Algorithm (Python AI Mirror)
Provides symmetric log-odds mapping between draft win probability and chess-style evaluation score.
"""

import math
from typing import Optional, Dict, Any

DEFAULT_SCALE = 1.5
DEFAULT_MAX_EVAL = 10.0
EPSILON = 1e-7


def format_eval_score(score: float) -> str:
    rounded = round(score, 1)
    if abs(rounded) < 1e-9:
        return "0.0"
    elif rounded > 0.0:
        return f"+{rounded:.1f}"
    else:
        return f"-{abs(rounded):.1f}"


def resolve_lead_category(score: float) -> str:
    if abs(score) < 0.05:
        return "EVEN"
    elif score > 1.5:
        return "BLUE_DECISIVE"
    elif score > 0.4:
        return "BLUE_ADVANTAGE"
    elif score > 0.0:
        return "SLIGHT_BLUE"
    elif score < -1.5:
        return "RED_DECISIVE"
    elif score < -0.4:
        return "RED_ADVANTAGE"
    else:
        return "SLIGHT_RED"


def winrate_to_eval_bar(
    blue_win_rate: float,
    scale: float = DEFAULT_SCALE,
    max_eval: float = DEFAULT_MAX_EVAL,
) -> Dict[str, Any]:
    """
    Converts blue win probability P into a symmetric log-odds Eval Bar score.
    """
    clamped_p = max(0.0, min(1.0, blue_win_rate))
    blue_pct = clamped_p * 100.0
    red_pct = 100.0 - blue_pct

    safe_p = max(EPSILON, min(1.0 - EPSILON, blue_win_rate))
    if abs(clamped_p - 0.50) < 1e-9:
        score = 0.0
    else:
        logit = math.log(safe_p / (1.0 - safe_p))
        score = max(-max_eval, min(max_eval, scale * logit))

    favored_side = "BLUE" if score > 1e-6 else ("RED" if score < -1e-6 else None)
    formatted = format_eval_score(score)
    category = resolve_lead_category(score)

    return {
        "score": round(score, 4),
        "formatted_score": formatted,
        "favored_side": favored_side,
        "blue_bar_percentage": round(blue_pct, 2),
        "red_bar_percentage": round(red_pct, 2),
        "lead_category": category,
    }


def eval_bar_to_winrate(score: float, scale: float = DEFAULT_SCALE) -> float:
    """
    Inverts an Eval Bar score back into blue win probability.
    """
    logit = score / scale
    return 1.0 / (1.0 + math.exp(-logit))

File: src/main/kotlin/com/loldraft/models/EvalBar.kt
package com.loldraft.models

import com.loldraft.data.models.Side
import kotlinx.serialization.Serializable
import java.util.Locale
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.round

@Serializable
enum class EvalLeadCategory {
    EVEN,
    SLIGHT_BLUE,
    BLUE_ADVANTAGE,
    BLUE_DECISIVE,
    SLIGHT_RED,
    RED_ADVANTAGE,
    RED_DECISIVE,
}

@Serializable
data class EvalBarScore(
    val score: Double,
    val formattedScore: String,
    val favoredSide: Side?,
    val blueBarPercentage: Double,
    val redBarPercentage: Double,
    val leadCategory: EvalLeadCategory,
)

object EvalBarCalculator {
    const val DEFAULT_SCALE: Double = 1.5
    const val DEFAULT_MAX_EVAL: Double = 10.0

    private const val EPSILON: Double = 1e-7

    fun calculate(
        blueWinRate: Double,
        scale: Double = DEFAULT_SCALE,
        maxEval: Double = DEFAULT_MAX_EVAL,
    ): EvalBarScore {
        val clampedP = blueWinRate.coerceIn(0.0, 1.0)
        val blueBarPercentage = clampedP * 100.0
        val redBarPercentage = 100.0 - blueBarPercentage

        val safeP = blueWinRate.coerceIn(EPSILON, 1.0 - EPSILON)
        val score =
            if (abs(clampedP - 0.50) < 1e-9) {
                0.0
            } else {
                val logit = ln(safeP / (1.0 - safeP))
                (scale * logit).coerceIn(-maxEval, maxEval)
            }

        val formatted = format(score)

        val favoredSide =
            when {
                score > 1e-6 -> Side.BLUE
                score < -1e-6 -> Side.RED
                else -> null
            }

        val leadCategory =
            when {
                abs(score) < 0.05 -> EvalLeadCategory.EVEN
                score > 1.5 -> EvalLeadCategory.BLUE_DECISIVE
                score > 0.4 -> EvalLeadCategory.BLUE_ADVANTAGE
                score > 0.0 -> EvalLeadCategory.SLIGHT_BLUE
                score < -1.5 -> EvalLeadCategory.RED_DECISIVE
                score < -0.4 -> EvalLeadCategory.RED_ADVANTAGE
                else -> EvalLeadCategory.SLIGHT_RED
            }

        return EvalBarScore(
            score = score,
            formattedScore = formatted,
            favoredSide = favoredSide,
            blueBarPercentage = blueBarPercentage,
            redBarPercentage = redBarPercentage,
            leadCategory = leadCategory,
        )
    }

    fun evalToWinRate(
        score: Double,
        scale: Double = DEFAULT_SCALE,
    ): Double {
        val logit = score / scale
        return 1.0 / (1.0 + exp(-logit))
    }

    fun format(score: Double): String {
        val rounded = round(score * 10.0) / 10.0
        return when {
            abs(rounded) < 1e-9 -> "0.0"
            rounded > 0.0 -> "+${String.format(Locale.US, "%.1f", rounded)}"
            else -> "-${String.format(Locale.US, "%.1f", abs(rounded))}"
        }
    }
}
